### PR TITLE
Adding support for OpenAI instance

### PR DESCRIPTION
**Pull request by carlotta94c**
## Purpose
This PR adds OpenAI API support for users who don't have access to Azure OpenAI. Additions:
- Calling OpenAI endpoint if env variable OPENAI_API_TYPE is set to openai (default models used: text-davinci-003 and gpt-3.5-turbo)
- Making bicep configuration conditional (Azure OpenAI resource is created only if param openAiType is set to 'azure')
- Adding instruction to switch from an Azure OpenAI endpoint (default) to a non-azure OpenAI instance
* ...

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[x] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone https://github.com/carlotta94c/azure-search-openai-demo
cd azure-search-openai-demo
git checkout openai_support_PR
npm install
```

* Test the code

Test deployment and app running in the cloud:

1. Navigate to the main.bicep file in the infra folder
2. Look for the openAiType parameter at line 16
3. Change its value to openai (instead of azure)
4. Configure the following app settings at line 135-136:
* OPENAI_API_ORGANIZATION={Your OpenAI organization}
* OPENAI_API_KEY={Your OpenAI API key}
5. Run azd up

To configure the local environment to use an existing OpenAI instance and test the app locally:

Run azd env set OPENAI_API_TYPE openai
Run azd env set OPENAI_API_ORGANIZATION {Your OpenAI organization}
Run azd env set OPENAI_API_KEY {Your OpenAI API key}

## What to Check
Verify that the following are valid
* Running 'azd up' without applying the changes to bicep file and env variables described above will ensure the same behavior of provisioning and deployment as before this PR (using Azure OpenAI resource is the default). 

## Other Information
Completion is not supported by OpenAI 'gpt-3.5-turbo'. For non-azure OpenAI instance the /chat endpoint is used.

### CODE REVIEW

File: app/backend/approaches/retrievethenread.py [hunk 1]
```diff
@@ -70,12 +71,22 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
...
         prompt = (overrides.get("prompt_template") or self.template).format(q=q, retrieved=content)
-        completion = openai.Completion.create(
-            engine=self.openai_deployment, 
-            prompt=prompt, 
-            temperature=overrides.get("temperature") or 0.3, 
-            max_tokens=1024, 
-            n=1, 
-            stop=["\n"])
+
+        if(self.openai_type == "azure"):
```
Comment (tonybaloney) [inline app/backend/approaches/retrievethenread.py:75]: the parenthesis aren't needed in Python

File: app/backend/approaches/retrievethenread.py [hunk 2]
```diff
@@ -70,12 +71,22 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
...
+            completion = openai.Completion.create(
+                engine=self.openai_deployment, 
+                prompt=prompt, 
+                temperature=overrides.get("temperature") or 0.3, 
+                max_tokens=1024, 
+                n=1, 
+                stop=["\n"])
+        else:
```
Comment (tonybaloney) [inline app/backend/approaches/retrievethenread.py:83]: These two conditions look almost identical? Is there a way to reduce duplication in this code

File: app/backend/approaches/readretrieveread.py [hunk 3]
```diff
@@ -90,7 +91,12 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
             prefix=overrides.get("prompt_template_prefix") or self.template_prefix,
             suffix=overrides.get("prompt_template_suffix") or self.template_suffix,
             input_variables = ["input", "agent_scratchpad"])
-        llm = AzureOpenAI(deployment_name=self.openai_deployment, temperature=overrides.get("temperature") or 0.3, openai_api_key=openai.api_key)
+        
+        if(self.openai_type == "azure"):
```
Comment (tonybaloney) [inline app/backend/approaches/readretrieveread.py:95]: The parenthesis aren't required

File: app/backend/approaches/readretrieveread.py [hunk 4]
```diff
@@ -90,7 +91,12 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
...
+            llm = AzureOpenAI(deployment_name=self.openai_deployment, temperature=overrides.get("temperature") or 0.3, openai_api_key=openai.api_key)
```
Comment (tonybaloney) [inline app/backend/approaches/readretrieveread.py:96]: `.get()` should take a the default argument as the second parameter, so you can replace `.get("temperature") or 0.3` with `get("temperature", 0.3)` in all occurrences ( I see a few in this PR )

File: app/backend/approaches/readretrieveread.py [hunk 5]
```diff
@@ -40,11 +40,12 @@ class ReadRetrieveReadApproach(Approach):
 
     CognitiveSearchToolDescription = "useful for searching the Microsoft employee benefits information such as healthcare plans, retirement plans, etc."
 
-    def __init__(self, search_client: SearchClient, openai_deployment: str, sourcepage_field: str, content_field: str):
+    def __init__(self, search_client: SearchClient, openai_deployment: str, openai_type, sourcepage_field: str, content_field: str):
```
Comment (tonybaloney) [inline app/backend/approaches/readretrieveread.py:43]: please add the type annotation as you have in the other constructor

File: app/backend/approaches/readdecomposeask.py [hunk 6]
```diff
@@ -68,7 +70,11 @@ def run(self, q: str, overrides: dict[str, Any]) -> Any:
         cb_handler = HtmlCallbackHandler()
         cb_manager = CallbackManager(handlers=[cb_handler])
 
-        llm = AzureOpenAI(deployment_name=self.openai_deployment, temperature=overrides.get("temperature") or 0.3, openai_api_key=openai.api_key)
+        if(self.openai_type == "azure"):
```
Comment (tonybaloney) [inline app/backend/approaches/readdecomposeask.py:73]: No parenthesis required

File: app/backend/approaches/chatreadretrieveread.py [hunk 7]
```diff
@@ -97,20 +98,48 @@ def run(self, history: Sequence[dict[str, str]], overrides: dict[str, Any]) -> A
...
-        completion = openai.Completion.create(
-            engine=self.chatgpt_deployment, 
-            prompt=prompt, 
-            temperature=overrides.get("temperature") or 0.7, 
-            max_tokens=1024, 
-            n=1, 
-            stop=["<|im_end|>", "<|im_start|>"])
-
-        return {"data_points": results, "answer": completion.choices[0].text, "thoughts": f"Searched for:<br>{q}<br><br>Prompt:<br>" + prompt.replace('\n', '<br>')}
+        if(self.openai_type == "azure"):
```
Comment (tonybaloney) [inline app/backend/approaches/chatreadretrieveread.py:101]: No parenthesis required

File: app/backend/app.py [hunk 8]
```diff
@@ -105,7 +124,8 @@ def ask():
     
 @app.route("/chat", methods=["POST"])
 def chat():
-    ensure_openai_token()
+    if(OPENAI_API_TYPE == "azure"):
```
Comment (tonybaloney) [inline app/backend/app.py:127]: No parenthesis

File: app/backend/app.py [hunk 9]
```diff
@@ -89,7 +106,9 @@ def content_file(path):
     
 @app.route("/ask", methods=["POST"])
 def ask():
-    ensure_openai_token()
+    if(OPENAI_API_TYPE == "azure"):
```
Comment (tonybaloney) [inline app/backend/app.py:109]: No parenthesis

File: app/backend/app.py [hunk 10]
```diff
@@ -52,16 +61,24 @@
     credential=azure_credential)
 blob_container = blob_client.get_container_client(AZURE_STORAGE_CONTAINER)
 
+if(OPENAI_API_TYPE == "azure"):
```
Comment (tonybaloney) [inline app/backend/app.py:64]: No parenthesis

File: README.md [hunk 11]
```diff
@@ -91,36 +95,68 @@ It will look like the following:
...
+3. Run `azd env refresh -e {environment name}` - Note that they will need the azd environment name, subscription Id, and location to run this command - you can find those values in your `./azure/{env name}/.env` file.  This will populate their azd environment's .env file with all the settings needed to run the app locally.
+4. Run `pwsh ./scripts/roles.ps1` - This will assign all of the necessary roles to the user so they can run the app locally.  If they do not have the necessary permission to create roles in the subscription, then you may need to run this script for them. Just be sure to set the `AZURE_PRINCIPAL_ID` environment variable in the azd .env file or in the active shell to their Azure Id, which they can get with `az account show`.
+
+### Switching from an Azure OpenAI endpoint to an OpenAI instance
+
+If you want to use an existing OpenAI instance, before running `azd up` to create your Azure resources and deploy the app on Azure, follow these steps:
+
+1. Navigate to the [main.bicep](./infra/main.bicep) file in the infra folder
+2. Look for the `openAiType` parameter at line 16
+3. Change its value to `openai` (instead of `azure`)
```
Comment (tonybaloney) [inline README.md:108]: AZD has parameters. We shouldn't require the user to manually change the bicep file. Alternatively, use the `azd set` command with environment variables

File: infra/main.bicep [hunk 12]
```diff
@@ -105,19 +115,30 @@ module backend 'core/host/appservice.bicep' = {
...
+      OPENAI_API_TYPE: openAiType
+    }:{
+      AZURE_STORAGE_ACCOUNT: storage.outputs.name
+      AZURE_STORAGE_CONTAINER: storageContainerName
+      AZURE_SEARCH_INDEX: searchIndexName
+      AZURE_SEARCH_SERVICE: searchService.outputs.name
+      OPENAI_API_TYPE: openAiType
+      OPENAI_GPT_MODEL: openAiGptModelName
+      OPENAI_CHATGPT_MODEL: openAiChatGptModelName
+      /*OPENAI_API_KEY: {your OpenAI key}
```
Comment (tonybaloney) [inline infra/main.bicep:135]: See comment about using parameters instead of editing bicep

File: notebooks/chat-read-retrieve-read.ipynb [hunk 13]
```diff
@@ -31,13 +37,19 @@
     "azure_credential = DefaultAzureCredential()\n",
     "\n",
     "# Used by the OpenAI SDK\n",
-    "openai.api_type = \"azure\"\n",
-    "openai.api_base = f\"https://{AZURE_OPENAI_SERVICE}.openai.azure.com\"\n",
-    "openai.api_version = \"2022-12-01\"\n",
+    "if(OPENAI_API_TYPE == \"azure\"):\n",
```
Comment (tonybaloney) [inline notebooks/chat-read-retrieve-read.ipynb:40]: Remove parenthesis

File: notebooks/chat-read-retrieve-read.ipynb [hunk 14]
```diff
@@ -101,12 +113,22 @@
...
     "\n",
     "if len(history) > 0:\n",
-    "    completion = openai.Completion.create(\n",
-    "        engine=AZURE_OPENAI_GPT_DEPLOYMENT,\n",
-    "        prompt=summary_prompt_template.format(summary=\"\\n\".join(history), question=user_input),\n",
-    "        temperature=0.7,\n",
-    "        max_tokens=32,\n",
-    "        stop=[\"\\n\"])\n",
+    "    \n",
+    "    if(OPENAI_API_TYPE == \"azure\"):\n",
```
Comment (tonybaloney) [inline notebooks/chat-read-retrieve-read.ipynb:117]: Remove parenthesis

File: notebooks/chat-read-retrieve-read.ipynb [hunk 15]
```diff
@@ -127,12 +149,20 @@
     "\n",
     "prompt = prompt_prefix.format(sources=content) + prompt_history + user_input + turn_suffix\n",
     "\n",
-    "completion = openai.Completion.create(\n",
-    "    engine=AZURE_OPENAI_CHATGPT_DEPLOYMENT, \n",
-    "    prompt=prompt, \n",
-    "    temperature=0.7, \n",
-    "    max_tokens=1024,\n",
-    "    stop=[\"<|im_end|>\", \"<|im_start|>\"])\n",
+    "if(OPENAI_API_TYPE == \"azure\"):\n",
```
Comment (tonybaloney) [inline notebooks/chat-read-retrieve-read.ipynb:152]: Remove parenthesis

**Review by tonybaloney — CHANGES_REQUESTED** [review]

**Comment by tonybaloney** [comment]
Thanks for the PR @carlotta94c this looks like a really useful feature. I've added some nitpick comments in the review, but the most important one is the use of parameters to set the OpenAI type instead of editing bicep files.

File: README.md (hunk 11 above)
Comment (mattmsft) [reply README.md:108]: Agreed. We may want to consider splitting it into a separate bicep module as well.

**Comment by carlotta94c** [comment]
> Thanks for the PR @carlotta94c this looks like a really useful feature. I've added some nitpick comments in the review, but the most important one is the use of parameters to set the OpenAI type instead of editing bicep files.

Hi @tonybaloney, huge thanks for your prompt review! I believe I've addressed the requested changes with my last commit.

File: app/backend/app.py [hunk 16]
```diff
@@ -33,14 +38,18 @@
 azure_credential = DefaultAzureCredential()
 
 # Used by the OpenAI SDK
-openai.api_type = "azure"
-openai.api_base = f"https://{AZURE_OPENAI_SERVICE}.openai.azure.com"
-openai.api_version = "2022-12-01"
+if(OPENAI_API_TYPE == "azure"):
```
Comment (tonybaloney) [inline app/backend/app.py:41]: Ah sorry one still left

**Review by tonybaloney — CHANGES_REQUESTED** [review]

File: app/backend/app.py (hunk 16 above)
Comment (carlotta94c) [reply app/backend/app.py]: Good catch, just removed!

**Review by tonybaloney — APPROVED** [review]

**Comment by carlotta94c** [comment]
closing this outdated PR as I opened a new one https://github.com/Azure-Samples/azure-search-openai-demo/pull/507

**Pull request closed without merging**